inst/interpolate.js

```diff
@@ -0,0 +1,315 @@
+// --- odin bits
+function interpolateCheckY(dimArg, dimTarget, nameArg, nameTarget) {
+    var rank = dimTarget.length - 1;
+    if (rank === 0) {
+        if (dimArg[0] !== dimTarget[0]) {
+            throw Error("Expected " + nameArg + " to have length " +
+                        dimArg[0] + " (for " + nameTarget + ")");
+        }
+    } else {
+        for (var i = 0; i < dimTarget.length; ++i) {
+            if (dimArg[i] !== dimTarget[i]) {
+                throw Error("Expected dimension " + (i + 1) + " of " +
+                            nameArg + " to have size " + dimArg[i] +
+                            " (for " + nameTarget + ")");
+            }
+        }
+    }
+}
+
+
+function interpolateCheckT(times, interpolateTimes, tcrit) {
+    if (times[0] < interpolateTimes.min) {
+        throw Error("Integration times do not span interpolation range; " +
+                    "min: " + interpolateTimes.min);
+    }
+    if (times[times.length - 1] > interpolateTimes.max) {
+        throw Error("Integration times do not span interpolation range; " +
+                    "max: " + interpolateTimes.max);
+    }
+    if (tcrit === undefined || tcrit === null) {
+        tcrit = interpolateTimes.max;
+    }
+    return tcrit;
+}
+
+
+// --- implementation
+function interpolateAlloc(type, x, y, failOnExtrapolate) {
+    var n = x.length;
+    var ny = y.length / n
+
+    var ret = {
+        type: type,
+        n: n,
+        ny: ny,
+        i: 0,
+        x: x.slice(),
+        failOnExtrapolate: failOnExtrapolate
+    };
+
+    // Optimisation for the constant interpolation - this should be
+    // moved into the C version of this too
+    if (type === "constant") {
+        var ylinear = y;
+        y = [];
+        for (var i = 0; i < n; ++i) {
+            var yi = [];
+            for (j = 0; j < ny; ++j) {
```

Comment:
I missed this, but there should be a `var j` here

Reply:
thanks - Ill add that into #7 in a minute :)

Reply:
OK, that's fixed on the other PR now